File: lib/ollama.ts
import { config } from "./config";

type OllamaEmbeddingResponse = {
  embedding?: number[];
  embeddings?: number[][];
};

type OllamaGenerateResponse = {
  response?: string;
};

export async function createEmbedding(input: string) {
  const response = await fetch(`${config.ollamaBaseUrl}/api/embeddings`, {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
    },
    body: JSON.stringify({
      model: config.ollamaEmbedModel,
      prompt: input,
    }),
  });

  if (!response.ok) {
    throw new Error(`Ollama embedding request failed: ${response.status}`);
  }

  const data = (await response.json()) as OllamaEmbeddingResponse;
  const embedding = data.embedding ?? data.embeddings?.[0];

  if (!embedding || embedding.length !== 768) {
    throw new Error(
      `Expected a 768-dimensional embedding, received ${embedding?.length ?? 0}`,
    );
  }

  return embedding;
}

export async function generateAnswer(prompt: string) {
  const controller = new AbortController();
  const timeout = setTimeout(() => controller.abort(), config.generationTimeoutMs);

  const response = await fetch(`${config.ollamaBaseUrl}/api/generate`, {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
    },
    signal: controller.signal,
    body: JSON.stringify({
      model: config.ollamaChatModel,
      prompt,
      stream: false,
      think: false,
      options: {
        num_predict: config.generationMaxTokens,
        temperature: 0.2,
      },
    }),
  }).finally(() => clearTimeout(timeout));

  if (!response.ok) {
    throw new Error(`Ollama generation request failed: ${response.status}`);
  }

  const data = (await response.json()) as OllamaGenerateResponse;
  const answer = data.response?.trim();

  if (!answer) {
    throw new Error("Ollama returned an empty answer.");
  }

  return answer;
}

File: backend/app/ollama.py
import httpx

from app.config import settings


def create_embedding(input_text: str) -> list[float]:
    response = httpx.post(
        f"{settings.ollama_base_url}/api/embeddings",
        json={
            "model": settings.ollama_embed_model,
            "prompt": input_text,
        },
        timeout=30,
    )
    response.raise_for_status()

    data = response.json()
    embedding = data.get("embedding")

    if embedding is None:
        embeddings = data.get("embeddings")
        if embeddings:
            embedding = embeddings[0]

    if not embedding or len(embedding) != settings.embedding_dimensions:
        received = len(embedding) if embedding else 0
        raise ValueError(
            f"Expected a {settings.embedding_dimensions}-dimensional embedding, "
            f"received {received}"
        )

    return embedding


def generate_answer(prompt: str) -> str:
    timeout_seconds = settings.generation_timeout_ms / 1000
    response = httpx.post(
        f"{settings.ollama_base_url}/api/generate",
        json={
            "model": settings.ollama_chat_model,
            "prompt": prompt,
            "stream": False,
            "think": False,
            "options": {
                "num_predict": settings.generation_max_tokens,
                "temperature": 0.2,
            },
        },
        timeout=timeout_seconds,
    )
    response.raise_for_status()

    answer = response.json().get("response", "").strip()

    if not answer:
        raise ValueError("Ollama returned an empty answer.")

    return answer
